File: Semester4/Systems for Design and Implementation/mpp-proiect-csharp-raulfrandes/networking/ClientWorker.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using model;
using model.DTOs;
using networking.dtos;
using services;

namespace networking;

public class ClientWorker : IObserver
{
    private IServices _server;
    private TcpClient _connection;

    private NetworkStream _stream;
    private IFormatter _formatter;
    private volatile bool _connected;
    private long _currentFilter;

    public ClientWorker(IServices server, TcpClient connection)
    {
        _server = server;
        _connection = connection;
        try
        {
            _stream = connection.GetStream();
            _formatter = new BinaryFormatter();
            _connected = true;
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }
    
    public virtual void Run()
    {
        while (_connected)
        {
            try
            {
                object request = _formatter.Deserialize(_stream);
                object response = HandleRequest((IRequest)request);
                if (response != null)
                {
                    SendResponse((IResponse)response);
                    if (response is ErrorResponse or LogoutResponse)
                    {
                        _stream.Flush();
                        _connected = false;
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }

            try
            {
                Thread.Sleep(5000);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        try
        {
            _stream.Close();
            _connection.Close();
        }
        catch (Exception e)
        {
            Console.WriteLine("Error " + e);
        }
    }
    
    public void PointsReceived(List<ParticipantDTO> participants)
    {
        Console.WriteLine("Points received");
        ParticipantDTO[] participantsArray = participants.ToArray();
        try
        {
            SendResponse(new ReceivePointsResponse(participantsArray));
        }
        catch (Exception e)
        {
            throw new ServiceException("Sending error: " + e);
        }
    }

    public void PointsReceivedFiltered(List<ParticipantDTO> participants)
    {
        Console.WriteLine("Points received filtered");
        ParticipantDTO[] participantsArray = participants.ToArray();
        try
        {
            SendResponse(new ReceivePointsFilteredResponse(participantsArray));
        }
        catch (Exception e)
        {
            throw new ServiceException("Sending error: " + e);
        }
    }

    public long GetFilter()
    {
        return _currentFilter;
    }

    private IResponse HandleRequest(IRequest request)
    {
        IResponse response = null;
        if (request is LoginRequest)
        {
            Console.WriteLine("Login request ...");
            LoginRequest loginRequest = (LoginRequest)request;
            string username = loginRequest.Username;
            try
            {
                lock (_server)
                { 
                    Referee referee = _server.Login(username, this);
                    return new LoginResponse(referee);
                }
            }
            catch (ServiceException e)
            {
                _connected = false;
                return new ErrorResponse(e.Message);
            }
        }

        if (request is LogoutRequest)
        {
            Console.WriteLine("Logout request");
            LogoutRequest logoutRequest = (LogoutRequest)request;
            long idReferee = logoutRequest.Id;
            try
            {
                lock (_server)
                {
                    _server.Logout(idReferee, this);
                }

                return new LogoutResponse();
            }
            catch (ServiceException e)
            {
                return new ErrorResponse(e.Message);
            }
        }

        if (request is GetParticipantsRequest)
        {
            Console.WriteLine("GetParticipants request");
            try
            {
                lock (_server)
                {
                    List<ParticipantDTO> participants = _server.GetParticipantsAndPoints();
                    ParticipantDTO[] participantsArray = participants.ToArray();
                    return new GetParticipantsResponse(participantsArray);
                }
            }
            catch (ServiceException e)
            {
                return new ErrorResponse(e.Message);
            }
        }

        if (request is GetTrialsRequest)
        {
            Console.WriteLine("GetTrials request");
            try
            {
                lock (_server)
                {
                    List<TrialDTO> trials = _server.GetTrialNames();
                    TrialDTO[] trialsArray = trials.ToArray();
                    return new GetTrialsResponse(trialsArray);
                }
            }
            catch (ServiceException e)
            {
                return new ErrorResponse(e.Message);
            }
        }

        if (request is AddResultRequest)
        {
            Console.WriteLine("AddResult request");
            AddResultRequest addResultRequest = (AddResultRequest)request;
            ResultDTO resultDto = addResultRequest.Result;
            Result result = DTOUtils.GetFromDTO(resultDto);
            try
            {
                lock (_server)
                {
                    _server.AddResult(result);
                    return new OkResponse();
                }
            }
            catch (ServiceException e)
            {
                return new ErrorResponse(e.Message);
            }
        }

        if (request is GetFilteredParticipantsRequest)
        {
            Console.WriteLine("GetFilteredParticipants request");
            GetFilteredParticipantsRequest getFilteredParticipantsRequest = (GetFilteredParticipantsRequest)request;
            long idTrial = getFilteredParticipantsRequest.IdTrial;
            try
            {
                lock (_server)
                {
                    List<ParticipantDTO> participants = _server.FilterParticipantsByTrial(idTrial);
                    ParticipantDTO[] participantsArray = participants.ToArray();
                    return new GetParticipantsResponse(participantsArray);
                }
            }
            catch (ServiceException e)
            {
                return new ErrorResponse(e.Message);
            }
        }

        if (request is SendFilterRequest)
        {
            Console.WriteLine("SendFilter request");
            SendFilterRequest sendFilterRequest = (SendFilterRequest)request;
            long idTrial = sendFilterRequest.IdTrial;
            try
            {
                _currentFilter = idTrial;
                return new OkResponse();
            }
            catch (ServiceException e)
            {
                return new ErrorResponse(e.Message);
            }
        }

        return response;
    }

    private void SendResponse(IResponse response)
    {
        Console.WriteLine("sending response " + response);
        lock (_stream)
        {
            _formatter.Serialize(_stream, response);
            _stream.Flush();
        }
    }
}

File: Semester4/Systems for Design and Implementation/mpp-proiect-java-raulfrandes/TriathlonNetworking/src/main/java/triathlon/networking/protocol/ClientWorker.java
package triathlon.networking.protocol;

import triathlon.model.DTOs.ParticipantDTO;
import triathlon.model.DTOs.TrialDTO;
import triathlon.model.Referee;
import triathlon.model.Result;
import triathlon.networking.protocol.dto.DTOUtils;
import triathlon.networking.protocol.dto.ResultDTO;
import triathlon.networking.protocol.requests.*;
import triathlon.networking.protocol.responses.*;
import triathlon.services.IObserver;
import triathlon.services.IServices;
import triathlon.services.ServiceException;

import java.io.EOFException;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;

public class ClientWorker implements Runnable, IObserver {
    private IServices server;
    private Socket connection;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;
    private volatile Long currentFilter;

    public ClientWorker(IServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        while (connected) {
            try {
                Object request = input.readObject();
                Object response = handleRequest((Request) request);
                if (response != null) {
                    sendResponse((Response) response);
                    if (response instanceof ErrorResponse) {
                        output.flush();
                        connected = false;
                    }
                    if (request instanceof LogoutRequest) {
                        output.flush();
                        connected = false;
                    }
                }
            } catch (EOFException e) {
                break;
            } catch (IOException e) {
                e.printStackTrace();
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error " + e);
        }
    }

    @Override
    public void pointsReceived(List<ParticipantDTO> participants) throws ServiceException {
        System.out.println("Points received");
        ParticipantDTO[] participantsArray = participants.toArray(ParticipantDTO[]::new);
        try {
            sendResponse(new ReceivePointsResponse(participantsArray));
        } catch (IOException e) {
            throw new ServiceException("Sending error: " + e);
        }
    }

    @Override
    public void pointsReceivedFiltered(List<ParticipantDTO> participants) throws ServiceException {
        System.out.println("Points received filtered");
        ParticipantDTO[] participantsArray = participants.toArray(ParticipantDTO[]::new);
        try {
            sendResponse(new ReceivePointsFilteredResponse(participantsArray));
        } catch (IOException e) {
            throw new ServiceException("Sending error: " + e);
        }
    }

    @Override
    public Long getFilter() {
        return currentFilter;
    }

    private Response handleRequest(Request request) {
        Response response = null;
        if (request instanceof LoginRequest loginRequest) {
            System.out.println("Login request ...");
            String username = loginRequest.getUsername();
            try {
                Referee referee = server.login(username, this);
                return new LoginResponse(referee);
            } catch (ServiceException e) {
                return new ErrorResponse(e.getMessage());
            }
        }
        if (request instanceof GetParticipantsRequest) {
            System.out.println("Get participants request");
            try {
                List<ParticipantDTO> participants = server.getParticipantsAndPoints();
                ParticipantDTO[] participantsArray = participants.toArray(ParticipantDTO[]::new);
                return new GetParticipantsResponse(participantsArray);
            } catch (ServiceException e) {
                return new ErrorResponse(e.getMessage());
            }
        }
        if (request instanceof GetTrialsRequest) {
            System.out.println("Get trials request");
            try {
                List<TrialDTO> trials = server.getTrialNames();
                TrialDTO[] trialsArray = trials.toArray(TrialDTO[]::new);
                return new GetTrialsResponse(trialsArray);
            } catch (ServiceException e) {
                return new ErrorResponse(e.getMessage());
            }
        }
        if (request instanceof AddResultRequest addResultRequest) {
            System.out.println("Add result request");
            ResultDTO resultDTO = addResultRequest.getResult();
            Result result = DTOUtils.getFromDTO(resultDTO);
            try {
                server.addResult(result);
                return new OkResponse();
            } catch (ServiceException e) {
                return new ErrorResponse(e.getMessage());
            }
        }
        if (request instanceof GetFilteredParticipantsRequest getFilteredParticipantsRequest) {
            System.out.println("Get filtered participants request");
            Long idTrial = getFilteredParticipantsRequest.getIdTrial();
            try {
                List<ParticipantDTO> participants = server.filterParticipantsByTrial(idTrial);
                ParticipantDTO[] participantsArray = participants.toArray(ParticipantDTO[]::new);
                return new GetParticipantsResponse(participantsArray);
            } catch (ServiceException e) {
                return new ErrorResponse(e.getMessage());
            }
        }
        if (request instanceof LogoutRequest logoutRequest) {
            Long idReferee = logoutRequest.getId();
            try {
                server.logout(idReferee, this);
                return new OkResponse();
            } catch (ServiceException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if (request instanceof SendFilterRequest sendFilterRequest) {
            Long idTrial = sendFilterRequest.getIdTrial();
            try {
                currentFilter = idTrial;
                return new OkResponse();
            } catch (ServiceException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        return response;
    }

    private void sendResponse(Response response) throws IOException {
        System.out.println("sending response " + response);
        synchronized (output) {
            output.writeObject(response);
            output.flush();
        }
    }
}
